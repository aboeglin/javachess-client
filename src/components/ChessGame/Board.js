import React from "react";
import { PropTypes } from "prop-types";
import {
  always,
  assoc,
  curry,
  findIndex,
  pipe,
  map,
  reverse,
  flatten,
  find,
  splitEvery,
  ifElse,
  identity
} from "ramda";
import { Square, SquareLabel } from "./styled";
import Piece from "./Piece";
import BOARD from "./board.json";
import withChess from "../../hoc/withChess";

// Move to separate file
const SquareContainer = ({ square, piece }) => (
  <Square
    color={square.color}
    x={square.x}
    y={square.y}
    key={`${square.x}${square.y}`}
    highlighted={square.highlighted}
  >
    <SquareLabel color={square.color}>
      {square.x}
      {square.y}
    </SquareLabel>
    {piece}
  </Square>
);

SquareContainer.propTypes = {
  square: PropTypes.object,
  piece: PropTypes.object
};

const orderSquaresForDisplay = pipe(
  reverse,
  splitEvery(8),
  map(reverse),
  flatten
);

const addPiece = curry((pieces, square) => ({
  square,
  piece: pieces.length ? getPieceAtSquare(square, pieces) : null
}));

const getPieceAtSquare = curry((square, pieces) =>
  pipe(
    find(({ x, y }) => x === square.x && y === square.y),
    ifElse(identity, piece => <Piece {...piece} />, always(null))
  )(pieces)
);

export const Board = ({ squares, pieces, possibleMoves }) =>
  pipe(
    map(
      ifElse(
        square =>
          findIndex(move => square.x === move.x && square.y === move.y)(
            possibleMoves
          ) > -1,
        assoc("highlighted", true),
        assoc("highlighted", false)
      )
    ),
    map(addPiece(pieces)),
    map(SquareContainer),
    orderSquaresForDisplay
  )(squares);

Board.propTypes = {
  squares: PropTypes.array,
  pieces: PropTypes.array,
  possibleMoves: PropTypes.array
};

Board.defaultProps = {
  squares: BOARD.squares,
  pieces: [],
  possibleMoves: []
};

export default withChess(Board);
