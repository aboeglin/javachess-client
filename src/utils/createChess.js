import SockJS from "sockjs-client";
import Stomp from "stompjs";

import { pipe, prop, append, forEach } from "ramda";

const HARDCODED_EMAIL = "user@domain.tld";

const createChess = () => {
  const socket = new SockJS("http://localhost:8080/ws");
  const ws = Stomp.over(socket);

  ws.debug = null;

  let observers = [];
  let gameId = null;

  let state = {
    pieces: [],
    possibleMoves: [],
    selection: null
  };

  ws.connect({}, () => {
    ws.subscribe("/user/queue/lfg/ack", frame => {
      gameId = pipe(prop("body"), JSON.parse, prop("gameId"))(frame);
      ws.subscribe(
        `/queue/game/${gameId}/ready`,
        pipe(prop("body"), body => {
          const newPieces = JSON.parse(body).game.board.pieces;
          updateState({ pieces: newPieces });
        })
      );

      ws.subscribe(
        `/queue/game/${gameId}/piece-moved`,
        pipe(prop("body"), body => {
          const newPieces = JSON.parse(body).game.board.pieces;

          // If move successful
          updateState({
            pieces: newPieces,
            possibleMoves: [],
            selection: null
          });
          // Else API should return MoveError and we should keep the possible moves
        })
      );

      ws.subscribe(
        `/queue/game/${gameId}/possible-moves`,
        pipe(prop("body"), body => {
          const possibleMoves = JSON.parse(body).possibleMoves;
          updateState({ possibleMoves: possibleMoves });
        })
      );

      ws.send(
        `/app/game/${gameId}/join`,
        {},
        JSON.stringify({ email: HARDCODED_EMAIL })
      );
    });

    ws.subscribe("/user/queue/errors", frame => {
      console.error(frame.body);
    });

    ws.send("/app/lfg", {}, JSON.stringify({ email: HARDCODED_EMAIL }));
  });

  const pieceSelected = piece => {
    updateState({ selection: piece });
    ws.send(
      `/app/game/${gameId}/select-piece`,
      {},
      JSON.stringify({ email: HARDCODED_EMAIL, x: piece.x, y: piece.y })
    );
  };

  const pieceMoved = (fromX, fromY, toX, toY) => {
    ws.send(
      `/app/game/${gameId}/perform-move`,
      {},
      JSON.stringify({ email: HARDCODED_EMAIL, fromX, fromY, toX, toY })
    );
  };

  const updateState = newState => {
    state = { ...state, ...newState };
    forEach(fn => fn(state))(observers);
  };

  // Store array of handlers, otherwise only the last one can get notified from state changed
  const onStateChanged = fn => {
    observers = append(fn, observers);
    fn(state);
  };

  return {
    onStateChanged,
    pieceSelected,
    pieceMoved
  };
};

export default createChess;
