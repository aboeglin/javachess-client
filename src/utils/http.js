import F from "fluture";
import axios from "axios";
import { curry } from "ramda";

export const post = curry((url, body) =>
  F((bad, good) => {
    axios
      .post(url, body)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  })
);

export const patch = curry((url, body) =>
  F((bad, good) => {
    axios
      .patch(url, body)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  })
);

export const get = (url) =>
  F((bad, good) => {
    axios
      .get(url)
      .then(good)
      .catch(bad);

    // Cancelation function
    return () => {};
  });
