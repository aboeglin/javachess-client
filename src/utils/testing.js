import React from "react";
import ReactDOM from "react-dom";
import { keys, pipe, propOr, curry } from "ramda";
import renderer from "react-test-renderer";
import cleanStack from "clean-stack";
import { Helmet } from "react-helmet";

beforeEach(() => {
  process.env.GATSBY_ENTRYPOINT = "/";
});

export const hasNonEmptyPropTypes = x => {
  const propTypes = propOr(false, "propTypes", x);
  // () => <span>cool</span> is valid, so no need for propTypes
  if (!propTypes) return true;
  // expect some keys if defined
  return pipe(keys, z => z.length > 0)(propTypes);
};

export const render = (C, props = {}, el = "div") => {
  try {
    const div = document.createElement(el);
    ReactDOM.render(
      <>
        <div id="leafletmap" />
        <C {...props} />
      </>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  } catch (error) {
    console.error(error.stack ? cleanStack(error.stack) : error);
  }
};
export const renderize = (C, props = {}, el = "div") => () => {
  try {
    const rend = () => {
      try {
        render(C, props, el);
      } catch (error) {
        if (error.stack) error.stack = cleanStack(error.stack);
        console.error(error.stack ? error.stack : error);
        throw error;
      }
    };
    expect(rend).not.toThrow();
  } catch (error) {
    console.error(error.stack ? cleanStack(error.stack) : error);
  }
};

const nameOf = C => C.displayName || C.name || "A Component";

export const renderWithTestRenderer = curry((C, props) => {
  it(`renders <${nameOf(C)} />`, () => {
    try {
      const c = renderer.create(<C {...props} />);
      let tree = c.toJSON();
      expect(tree).toMatchSnapshot();
    } catch (error) {
      console.log(error.stack ? cleanStack(error.stack) : error);
    }
  });
});

export const standardComponentTest = (C, props = {}, el = "div") => {
  renderWithTestRenderer(C, props);
  it("renders without crashing", renderize(C, props, el));
  it("has propTypes", () => {
    expect(hasNonEmptyPropTypes(C)).toBeTruthy();
  });
};

export const standardComponentTestWithHeadTest = (
  C,
  props = {},
  el = "div",
  headProperty = "",
  headData = null
) => {
  renderWithTestRenderer(C, props);
  it("renders without crashing", renderize(C, props, el));
  it("has propTypes", () => {
    expect(hasNonEmptyPropTypes(C)).toBeTruthy();
  });
  it("has correct head props", () => {
    headData
      ? expect(Helmet.peek()).toHaveProperty(headProperty, headData)
      : expect(Helmet.peek()).toHaveProperty(headProperty);
  });
};
