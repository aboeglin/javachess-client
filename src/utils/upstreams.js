import { __, concat, curry, map, pipe, prop, toString } from "ramda";

import { get, patch, post } from "./http";

const ENDPOINT =
  process.env.GATSBY_CHESS_SERVICE_ENDPOINT || "http://localhost:8080";
const GET_GAMES_PATH = "/games";
const POST_GAMES_PATH = "/games";
const PATCH_GAMES_PATH = "/games";

export const fetchGamesWithGetAndEndpoint = curry((get, endpoint, path) =>
  pipe(concat(endpoint), get, map(prop("data")))(path)
);

export const fetchGames = () =>
  fetchGamesWithGetAndEndpoint(get, ENDPOINT, GET_GAMES_PATH);

export const createGameWithPostAndEndpoint = curry(
  (post, endpoint, path, userName) =>
    pipe(
      concat(endpoint),
      post(__, { userId: userName }),
      map(prop("data"))
    )(path)
);

export const createGame = createGameWithPostAndEndpoint(
  post,
  ENDPOINT,
  POST_GAMES_PATH
);

export const joinGameWithPatchAndEndpoint = curry(
  (patch, endpoint, path, userName, gameId) =>
    pipe(
			toString,
      concat("/"),
      concat(path),
      concat(endpoint),
      patch(__, { userId: userName }),
      map(prop("data"))
    )(gameId)
);

export const joinGame = joinGameWithPatchAndEndpoint(patch, ENDPOINT, PATCH_GAMES_PATH);
