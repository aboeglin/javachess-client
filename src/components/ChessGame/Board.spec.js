import { Board } from "./Board";
import { standardComponentTest } from "../../utils/testing";

standardComponentTest(Board, {
  possibleMoves: [{ x: "b", y: "3" }],
  pieces: [] // Add it back when figured a way to handle injection of chess provider
  // pieces: [{ type: "PAWN", color: "WHITE", x: "b", y: "2" }]
});
