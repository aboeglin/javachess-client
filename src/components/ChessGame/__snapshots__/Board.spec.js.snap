// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <Board /> 1`] = `
Array [
  <div
    className="css-1495z7x"
    color="WHITE"
    x="a"
    y="8"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      a
      8
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="b"
    y="8"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      b
      8
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="c"
    y="8"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      c
      8
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="d"
    y="8"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      d
      8
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="e"
    y="8"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      e
      8
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="f"
    y="8"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      f
      8
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="g"
    y="8"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      g
      8
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="h"
    y="8"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      h
      8
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="a"
    y="7"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      a
      7
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="b"
    y="7"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      b
      7
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="c"
    y="7"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      c
      7
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="d"
    y="7"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      d
      7
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="e"
    y="7"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      e
      7
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="f"
    y="7"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      f
      7
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="g"
    y="7"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      g
      7
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="h"
    y="7"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      h
      7
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="a"
    y="6"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      a
      6
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="b"
    y="6"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      b
      6
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="c"
    y="6"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      c
      6
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="d"
    y="6"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      d
      6
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="e"
    y="6"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      e
      6
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="f"
    y="6"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      f
      6
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="g"
    y="6"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      g
      6
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="h"
    y="6"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      h
      6
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="a"
    y="5"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      a
      5
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="b"
    y="5"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      b
      5
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="c"
    y="5"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      c
      5
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="d"
    y="5"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      d
      5
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="e"
    y="5"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      e
      5
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="f"
    y="5"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      f
      5
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="g"
    y="5"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      g
      5
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="h"
    y="5"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      h
      5
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="a"
    y="4"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      a
      4
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="b"
    y="4"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      b
      4
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="c"
    y="4"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      c
      4
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="d"
    y="4"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      d
      4
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="e"
    y="4"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      e
      4
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="f"
    y="4"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      f
      4
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="g"
    y="4"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      g
      4
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="h"
    y="4"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      h
      4
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="a"
    y="3"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      a
      3
    </div>
  </div>,
  <div
    className="css-6ks6ln"
    color="WHITE"
    x="b"
    y="3"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      b
      3
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="c"
    y="3"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      c
      3
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="d"
    y="3"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      d
      3
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="e"
    y="3"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      e
      3
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="f"
    y="3"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      f
      3
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="g"
    y="3"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      g
      3
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="h"
    y="3"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      h
      3
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="a"
    y="2"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      a
      2
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="b"
    y="2"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      b
      2
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="c"
    y="2"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      c
      2
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="d"
    y="2"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      d
      2
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="e"
    y="2"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      e
      2
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="f"
    y="2"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      f
      2
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="g"
    y="2"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      g
      2
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="h"
    y="2"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      h
      2
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="a"
    y="1"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      a
      1
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="b"
    y="1"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      b
      1
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="c"
    y="1"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      c
      1
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="d"
    y="1"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      d
      1
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="e"
    y="1"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      e
      1
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="f"
    y="1"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      f
      1
    </div>
  </div>,
  <div
    className="css-15k27lr"
    color="BLACK"
    x="g"
    y="1"
  >
    <div
      className="css-1y32jx2"
      color="BLACK"
    >
      g
      1
    </div>
  </div>,
  <div
    className="css-1495z7x"
    color="WHITE"
    x="h"
    y="1"
  >
    <div
      className="css-1nwny1d"
      color="WHITE"
    >
      h
      1
    </div>
  </div>,
]
`;
